test/assignment-5/CollateralizedVault.t.sol

```diff
@@ -0,0 +1,270 @@
+
+        assertEq(to, (12345 * 1e18) / 1e3);
+
+        // test reverse
+        assertEq(vault.scaleInteger(to, 18, 6), from);
+    }
+}
+
+abstract contract DepositedCollateralState is ZeroState {
+    function setUp() public virtual override {
+        super.setUp();
+
+        vm.prank(USER);
+        vault.deposit(3 ether);
+
+        vm.prank(USER);
+        dai.approve(address(vault), type(uint256).max);
+    }
+}
+
+contract DepositedCollateralStateTest is DepositedCollateralState {
+    function testBorrow() public {
+        vm.prank(USER);
+        vault.borrow(6000 * 1e18);
+
+        // 6000 DAI was transfered to the USER
+        assertEqDecimal(dai.balanceOf(USER), 6000 * 1e18, 18);
+
+        // User debt is recorded
+        assertEq(vault.borrows(USER), 6000 * 1e18);
+    }
+
+    function testBorrowMultipleTimesAccumulatesDebt() public {
+        vm.startPrank(USER);
+
+        vault.borrow(5 * 1e18);
+        vault.borrow(10 * 1e18);
+
+        assertEqDecimal(dai.balanceOf(USER), 15 * 1e18, 18);
+        assertEq(vault.borrows(USER), 15 * 1e18);
+    }
+
+    function testRevertsWhenTryingToBorrowTooMuch() public {
+        vm.prank(USER);
+        vm.expectRevert(CollateralizedVault.NotEnoughCollateral.selector);
+        vault.borrow(6000 * 1e18 + 1);
+    }
+}
+
+abstract contract BorrowedState is DepositedCollateralState {
+    function setUp() public virtual override {
+        super.setUp();
+
+        vm.prank(USER);
+        vault.borrow(6000 * 1e18);
+    }
+}
+
+contract BorrowedStateTest is BorrowedState {
+    function testRepayDebt() public {
+        assertEq(vault.borrows(USER), 3 * 2000 ether);
+
+        vm.prank(USER);
+        vault.repay(2000 ether);
+
+        assertEq(vault.borrows(USER), 2 * 2000 ether);
+        assertEq(dai.balanceOf(USER), 2 * 2000 ether);
+    }
+
+    function testRepayTooMuchDebtReverts() public {
+        vm.prank(USER);
+        vm.expectRevert(stdError.arithmeticError);
+        vault.repay(6001 ether);
+    }
+
+    function testCantWithdrawCollateral() public {
+        vm.prank(USER);
+        vm.expectRevert(CollateralizedVault.TooMuchDebt.selector);
+        vault.withdraw(1);
+    }
+
+    function testCanWithdrawEntireCollateralIfPaidAllDebt() public {
+        assertEq(weth.balanceOf(USER), 7 ether);
+        assertEq(vault.borrows(USER), 6000 * 1e18);
+        assertEq(vault.deposits(USER), 3 ether);
+
+        // Repay entire debt
+        vm.startPrank(USER);
+        vault.repay(3 * 2000 ether);
+        // No more debt
+        assertEq(vault.borrows(USER), 0);
+        // But also no more DAI
+        assertEq(dai.balanceOf(USER), 0);
+
+        vault.withdraw(3 ether);
+        // WETH is returned to user
+        assertEq(weth.balanceOf(USER), 10 ether);
+    }
+
+    function testOnlyOwnerCanLiquidate() public {
+        vm.prank(address(0x1234));
+        vm.expectRevert("Ownable: caller is not the owner");
```

Comment:
good test. 

this is a situation where we should always test "outside functionality" because what if u forgot to add the modifier!